File: ffshark_send_packets.py
import subprocess
import os
import argparse
import string
import random
import sys
import glob
import time
import math
import fcntl

# import libmpsoc provided by Clark
# make sure it's installed as python module by running sudo pip3 install libmpsoc
# should already be installed but incase it isn't refer to above comment
import libmpsoc as mp

# The following script randomly selects packets from a directory containing packet text files to ffshark. It will check if the fifo has
# enough space for txt file packet.

# Register constants
# The base addr can vary depending on FFShark and PS config, we really shouldn't hard code
FFSHARK_BASE = 0xA0010000
FFSHARK_SIZE = 0x1000 # 4KB mem map region
FFSHARK_ENABLE_OFFSET = 0x4
FIFO_BASE = 0xA0011000
FIFO_SIZE = 0x1000 # 4KB mem map region
FIFO_TDFV_OFFSET = 0xC
FIFO_TDFD_OFFSET = 0x10
FIFO_TLR_OFFSET = 0x14
FIFO_SRR_OFFSET = 0x28
FIFO_SRR_RST_VAL = 0xA5
FIFO_DATA_WIDTH = 4 #in bytes

# Lock file used to maintain thread safety for multiple
# processes running on the FPGA
LOCK_FILE = "/var/lock/pokelockfile"

def init_lock_file():
	if not os.path.exists(LOCK_FILE):
		os.system("touch " + LOCK_FILE)

def main():
    parser = argparse.ArgumentParser(description="Sends packets randomly one by one to ffshark when given packet text files directory")
    parser.add_argument("--packets-directory", action="store", help="Provide directory of packet text files", required="true")
    parser.add_argument("--send-wait-time", action="store", type=float, default=0, help="Specify wait time in seconds between sending packets.")
    parser.add_argument("--num-packets", action="store", type=float, default=float("inf"), help="Specify how many random packets to send.")
    parser.add_argument("--debug", action="store_true", help="Enable for more verbosity")
    parser.add_argument("--perf-test", action="store_true", help="measure performance")
    args = parser.parse_args()

    directory = args.packets_directory
    wait_time = args.send_wait_time
    num_packets = args.num_packets
    perf_test = args.perf_test

    directory = args.packets_directory
    assert(os.path.exists(directory)), "directory doesn't exist"
    directory = os.path.join(directory, '')
    packet_files_list = glob.glob(directory + "*.txt")

    # initialize the lock
    init_lock_file()
    lock = open(LOCK_FILE, "r")

    fcntl.flock(lock.fileno(), fcntl.LOCK_EX)

    # init the axi lite mem map for FFShark and FIFO
    axil_FFShark = mp.axilite(addr=FFSHARK_BASE,size=FFSHARK_SIZE)
    axil_FIFO = mp.axilite(addr=FIFO_BASE,size=FIFO_SIZE)

    # initialize FFShark and FIFO
    print(axil_FFShark.write32(value=0x1,offset=FFSHARK_ENABLE_OFFSET))
    print(axil_FIFO.write32(value=FIFO_SRR_RST_VAL,offset=FIFO_SRR_OFFSET))

    fcntl.flock(lock.fileno(), fcntl.LOCK_UN)

    if (perf_test):
        start_time = time.time()
        total_bytes = 0
        total_write_time = 0
        file_read_time = 0
        write_word_no_lock_time = 0

    # sending packets
    iteration_count = 0
    while (iteration_count < num_packets):
        # pick a random file
        packet_file_index = random.randint(0, len(packet_files_list) - 1)
        file = packet_files_list[packet_file_index]
        # get file size and round up to nearest word
        size_bytes = os.path.getsize(file)
        
        if (perf_test):
            total_bytes += size_bytes            
        
        # divide by 8 because each hex char is UTF-8 so 1 byte but it represents only 4 bits so it's 2x the data
        num_words = int(math.ceil(size_bytes/8))
        if (args.debug):
            print(num_words)
            print(size_bytes)
        fcntl.flock(lock.fileno(), fcntl.LOCK_EX)
        num_vacant_words = axil_FIFO.read32(offset=FIFO_TDFV_OFFSET)
        fcntl.flock(lock.fileno(), fcntl.LOCK_UN)
        if (args.debug):
            print("num vacant words ::: " + str(num_vacant_words))
        
        # check if the FIFO has enough space for the packet and if not wait
        while (num_words > num_vacant_words):
            fcntl.flock(lock.fileno(), fcntl.LOCK_EX)
            num_vacant_words = axil_FIFO.read32(offset=FIFO_TDFV_OFFSET)
            fcntl.flock(lock.fileno(), fcntl.LOCK_UN)
            print("num vacant words ::: " + str(num_vacant_words))
        
        # write the packet a word at a time
        if (perf_test):
            write_time = time.time()
        fcntl.flock(lock.fileno(), fcntl.LOCK_EX)
        with open(file) as f:
            for i in range(num_words):
                if (perf_test):
                    file_time = time.time()
                file_val = int(f.read(8),16)
                if (perf_test):
                    file_read_time += time.time() - file_time               
                
                if (perf_test):
                    write_no_lock_time = time.time()
                axil_out = axil_FIFO.write32(value=file_val,offset=FIFO_TDFD_OFFSET)
                if (perf_test):
                    write_word_no_lock_time += time.time() - write_no_lock_time                
                
                if (args.debug):
                    print(hex(file_val))
                    print(axil_out)
        
        # Write the number of bytes to TLR
        # divide by 2 because each hex char is UTF-8 so 1 byte but it represents only 4 bits so it's 2x the data
        axil_out = axil_FIFO.write32(value=int(math.ceil(size_bytes/2)),offset=FIFO_TLR_OFFSET)
        if (args.debug):
            print(int(math.ceil(size_bytes/2)))
            print(axil_out)
        fcntl.flock(lock.fileno(), fcntl.LOCK_UN)
        if (perf_test):
            total_write_time += time.time() - write_time
            
        
        if (wait_time > 0):
            time.sleep(wait_time)
        iteration_count += 1
        
    if (perf_test):
        total_time = time.time() - start_time
        bit_rate = (total_bytes * 8) / total_time
        write_bit_rate = (total_bytes * 8) / total_write_time
        print("Total time : " + str(total_time))
        print("Write time : " + str(total_write_time))
        print("Write with no Locking time: " + str(write_word_no_lock_time))
        print("File read time: " + str(file_read_time))
        print("Total Data rate : " + str(bit_rate) + " bits/second")


if __name__ == "__main__":
    main()

File: ffshark_c_lang_drivers/ffshark_send_packets.c
#include "libmpsoc.h"
#include <glob.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <time.h>
#include <inttypes.h>
#include <getopt.h>
#include <errno.h>
#include <math.h>

// Register constants
// The base addr can vary depending on FFShark and PS config, we really shouldn't hard code
#define FFSHARK_BASE 0xA0010000
#define FFSHARK_SIZE 0x1000 // 4KB mem map region
#define FFSHARK_ENABLE_OFFSET 0x4
#define FIFO_BASE 0xA0011000
#define FIFO_SIZE 0x1000 // 4KB mem map region
#define FIFO_TDFV_OFFSET 0xC
#define FIFO_TDFD_OFFSET 0x10
#define FIFO_TLR_OFFSET 0x14
#define FIFO_SRR_OFFSET 0x28
#define FIFO_SRR_RST_VAL 0xA5
#define FIFO_DATA_WIDTH 4 //in bytes

void print_help(){
    printf("usage: ffshark_send_packets [-n NUM] [-d DIRECTORY] [-r NUM] [-h]\n\n"
        "send pregenerated packets from text files to FFSHark.\n\n"
        "Optional Arguments:\n"
        "  -h, --help\n"
        "          Show this help message and exit\n"
        "  -n, --num-packets=NUM\n"
        "          The number of packets to send.\n"
        "          A value of -1 or less is equivalent to infinity. It is -1 by default\n"
        "  -r, --reset-fifo=NUM\n"
        "          Set it to any number besides 0 to reset the AXI send fifo. Set it to 0 to disable reset.\n" 
        "          Reset is disabled by default\n"
        "Mandatory Arguments:\n"
        "  -d, --directory=DIRECTORY\n"
        "          This is the directory of pregenerated test packets. These packets must be in .txt files \n"
);
    return;

}

int main(int argc, char **argv) {
    //Get user arguments
    int opt;
    int num_packets = -1;
    char *directory = NULL;
    char directory_pattern[1024];
    char *endptr;
    int option_index;
    int reset_fifo = 0;
    struct option long_options[] = {
        {"help", no_argument, NULL, 'h'},
        {"directory", required_argument, NULL, 'd'},
        {"num-packets", required_argument, NULL, 'n'},
        {"reset-fifo", required_argument, NULL, 'r'},
        {0, 0, 0, 0}
    };

    while ((opt = getopt_long(argc, argv, "n:f:h", long_options, &option_index)) != -1){
        switch(opt){
            case 'n':
                num_packets = strtol(optarg, &endptr, 10);
                if (endptr == optarg || *endptr != '\0'){
                    printf("ERROR: -n or --num-packets argument was not a valid integer.\nExiting.\n");
                    return 1;
                } else if (errno == ERANGE){
                    printf("ERROR: -n or --num-packets argument was too large and overflowed.\n"
                        "Try reducing NUM or setting to -1 for infinity.\nExiting\n");
                    return 1;
                }
                break;
            case 'd':
                directory = optarg;
                break;
            case 'r':
                reset_fifo = strtol(optarg, &endptr, 10);
                if (endptr == optarg || *endptr != '\0'){
                    printf("ERROR: -r or --reset-fifo argument was not a valid integer.\nExiting.\n");
                    return 1;
                } else if (errno == ERANGE){
                    printf("ERROR: -r or --reset-fifo argument was too large and overflowed.\n"
                        "to disable fifo --reset-fifo needs to be 0. To turn it on, --reset-fifo can be any other number.\n" 
                        "It does not have to be so large.\nExiting\n");
                    return 1;
                }
                break;
            case 'h':
                print_help();
                return 0;
            default:
                print_help();
                return 1;
        }
    }
  
    
    //get the directory pattern
    if (directory == NULL){
        printf ("You did not provide directory of packets");
        return -1;
    }
    strcpy(directory_pattern, directory);
    if (directory_pattern[strlen(directory_pattern)-1] == '/'){
        strcat(directory_pattern, "*.txt");
    }
    else {
        strcat(directory_pattern, "/*.txt");
    }
    
    //initialize locking mechanism
    lock_init();
    
    int lock_fd = lock();    
    //initialize axi lite mem maps for FFShark and FIFO
    AXILITE axil_ffshark;
    AXILITE axil_fifo;
    init_axilite(&axil_fifo, FIFO_BASE, FIFO_SIZE);
    init_axilite(&axil_ffshark, FFSHARK_BASE, FFSHARK_SIZE);
    
    //init FFShark and FIFO
    write_axilite(&axil_ffshark, FFSHARK_ENABLE_OFFSET, 0x1);
    if (reset_fifo != 0)
        write_axilite(&axil_fifo, FIFO_SRR_OFFSET, FIFO_SRR_RST_VAL);
    unlock(lock_fd);
    
    //get all the packet text files in the given directory
    glob_t file_names;
    int glob_res;    
    glob_res =  glob(directory_pattern, 0, NULL, &file_names);
    if (glob_res == GLOB_NOMATCH || glob_res == GLOB_ABORTED || glob_res == GLOB_NOSPACE){
        printf("Could not find matching pattern for files\n");
        printf("Either directory does not exist, or the directory does not have .txt files for its packets\n");
        return -1;
    }
    
    FILE *fp;
    struct stat st;
    int iteration_count = 0;
    int packet_file_index;
    unsigned num_vacant_words, num_words;
    unsigned packet_data;
    unsigned char word_buffer[10];
    word_buffer[0] = '0';
    word_buffer[1] = 'x';
        
    srand(time(NULL));

    //for profiling    
    double total_bytes = 0;
    double file_read_time = 0;
    double total_write_time = 0;
    double write_word_no_lock_time = 0;
    clock_t start_time = clock();
    
    while(iteration_count < num_packets || num_packets <= -1){
        
        //choose random packet file 
        packet_file_index = rand() % file_names.gl_pathc;
        if(stat(file_names.gl_pathv[packet_file_index], &st) != 0) {
            printf("invalid packet file");
            return -1;
        }
        
        //divide by 8 because each hex char is UTF-8 so 1 byte but it represents only 4 bits so it's 2x the data
        num_words = st.st_size / 8;
        total_bytes += st.st_size / 2;        
        
        //check if FIFO has enough space for packet
        lock_fd = lock();
        num_vacant_words = read_axilite(&axil_fifo, FIFO_TDFV_OFFSET);
        unlock(lock_fd);
        while (num_words - 8 > num_vacant_words){
            lock_fd = lock();
            num_vacant_words = read_axilite(&axil_fifo, FIFO_TDFV_OFFSET);
            unlock(lock_fd);
            printf("num vacant words: %u\n", num_vacant_words);
        }
        
        //write the packet one word at a time
        fp = fopen(file_names.gl_pathv[packet_file_index], "r");
        if (fp == NULL){
            printf("file open failed");
            return -1;
        }
        
        clock_t write_start = clock();
        lock_fd = lock();
        while (fread(&word_buffer[2], 8, 1, fp) > 0){
            //profile the amount of time to read out packet file
            clock_t file_start_time = clock();
            packet_data = strtoul((char *)word_buffer, NULL, 0);
            file_read_time += (double) (clock() - file_start_time)/(CLOCKS_PER_SEC/1000000);
            
            //profile register writing time
            clock_t write_word_start = clock();
            write_axilite(&axil_fifo, FIFO_TDFD_OFFSET, packet_data);
            write_word_no_lock_time += (clock() - write_word_start)/(CLOCKS_PER_SEC/1000000);
        }
        fclose(fp);        
        
        //Write the number of bytes to TLR
        //divide by 2 because each hex char is UTF-8 so 1 byte but it represents only 4 bits so it's 2x the data
        write_axilite(&axil_fifo, FIFO_TLR_OFFSET, (unsigned)(st.st_size / 2));
        unlock(lock_fd);        
        total_write_time += (double)(clock() - write_start)/(CLOCKS_PER_SEC/1000000);
        
        iteration_count+=1;
        //printf("iteration count: %d\n", iteration_count);
        
    }
    
    //print out profiling data
    clock_t end_time = clock();    
    double total_time = (double) (end_time - start_time)/(CLOCKS_PER_SEC/1000000);
    double bit_rate = ((total_bytes * 8)/total_time) * 1000000;
    printf("Total time : %f seconds\n", total_time / 1000000 );
    printf("Write time : %f seconds\n ", total_write_time / 1000000);
    printf("Write with no locking time : %f seconds\n ", write_word_no_lock_time / 1000000);
    printf("File read time: %f seconds\n", file_read_time / 1000000);
    printf("Bit rate : %f bits per second \n", bit_rate);

}